### ISSUE TITLE
Question: What sort of speeds should I be getting?

### ISSUE THREAD

**Issue by Pluckerpluck** [xlwings/xlwings#1086]
#### OS (e.g. Windows 10 or macOS Sierra)

Windows 10

#### Versions of xlwings, Excel and Python (e.g. 0.11.8, Office 365, Python 3.7)

0.15.2, Python 3.6, Office Pro 2010

#### Describe your issue

So I was wondering what type of speeds I should be getting when writing to a workbook from within a macro, because I feel like my script is running fairly slowly.

I currently extract a bunch of data from a worksheet (I pull in the entire worksheet and then process it), before outputting a variety of tables. Currently, for testing purposes, I output 6 of these tables, each 11x40.

These tables are stored as pandas objects, and I currently try to export them using:

```sheet.range("A1").value = result_dataframe```

The thing is, this takes over 200ms to run. Is that about normal? As a more static reference, running:

```workbook.app.screen_updating = False```

takes ~10ms and listing all the sheet names:

`sheets = {sheet.name for sheet in workbook.sheets}`

takes 65ms.

---------

It just feels kind of long time given that it's not that much detail. I want to paste this on multiple different sheets as well, but there doesn't seem to be a way to batch that operation which means this can quite quickly add up if I'm not careful with how I output my data. 

Not a terribly large amount, and a wait we can probably live with, but my program takes 0.5 seconds to run in VBA. The python code implementation takes something like 0.1-0.2 seconds, but then I have to spend 3-5 seconds just pasting the data back.

Given that pages [like this](https://www.pyxll.com/blog/performance-comparison/) suggest that ExcelPython could write 1000 cells in 0.026491s, and I have a not too shabby 7th gen laptop i5 (3.2Ghz) I feel like something is burning up some time here.

Or is this just the way it is? Reading an **entire worksheet** only takes 25ms, yet writing a few hundred cells seems to take 10x as long!

----------

Just to stress why this is a problem. There is a possibility of me wanting to write this 400 cell dataframe 5 times on a page, with 6 variants (i.e. sheets) for a total of 30 rights. That's over 7 seconds of loading time for something VBA does near instantly. And technically it needs to support up to 30 per page! Which would be 45 seconds. VBA runs this 0.5 seconds (using a much less efficient method)

As a result I'm trying to avoid writing the 5 individual frames, instead only writing the **entire page** but this is chaotic as you have to create very weird structures to achieve this. It's not nice to use at all (plus I still have to do it per sheet)

So even from a usability standpoint, performance is highly important.

----------

Extra information: It appears writes of any size (at least coming from Pandas) takes ~245ms to run. Just tried it on a table of 7x10.

**Comment by fzumstein** [comment]
Generally speaking, xlwings has been optimized for user friendliness, not speed. Having said that, here's what we we're looking at:

* every cross app call adds overhead, that's why you should always write arrays instead of writing cell by cell (you're doing that already by sending over whole dataframes)
* xlwings does a lot of slow cell by cell data cleaning by default. To skip that, have a look at:  http://docs.xlwings.org/en/stable/api.html#xlwings.Range.raw_value

**Comment by Pluckerpluck** [comment]
My issue is more the discrepancy between reads and rights.

I would have expected them to have similar call overheads, and yet a write takes over 250ms, while a read takes 25...

I'll investigate raw as that may help, but the write time seems independent of data frame size, so I'm unsure.

------

Have you considered the concept of a "cached output/input sheet"?

A limited functionality sheet (only data reads and writes) that uses a python context manager to bulk write values in one go. It's looking like I'm going to have to create one for my purposes, so I was wondering if it had come up before (as this could still be user friendly if done correctly)

**Comment by Pluckerpluck** [comment]
Raw helped **massively**. Turns out writing a Pandas dataframe to Excel has a 220ms hit on my laptop (on top of the COM server ~30ms) **regardless** of size. That adds up super quickly when you want to plant down even just 10 tables and have to wait 2.5 seconds.

So I created my own `BufferedInputSheet` and `BufferedOutputSheet`. They work as a context manager and mimic xlwings data input:

```python
with BufferedInputSheet(xlwings_sheet) as buffer:
    data = buffer.range("A1").value
```

but work by loading the entire sheet into a Pandas dataframe. `BufferedOutputSheet` is instead backed by a raw Numpy array to avoid huge write speed issues when writing Pandas to Excel. 

Given that I rely on Pandas and Numpy (and the fact that the code style isn't even close to matching xlwings) these aren't suitable for the public repo, but I believe the concept is a good one.

-----

It's not simple to implement of course, but I thought I'd throw down the suggestion before closing this issue.

The idea being:

BufferedSheet() -> Loads contents of a sheet into python 2D array. Store changes in the 2D array. Write it back when you leave the context.

And it taking options like `override` which would skip the initial read entirely.

Thus is would both be performant and user accessible.
